### PR TITLE
Feat: fetchCurrentDevice API Implementation

### PR DESCRIPTION
**Pull request by hahnandrew**
*Description of changes:*

Implemented `fetchCurrentDevice` API

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by khatruong2009 — APPROVED** [review]

File: packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart [hunk 1]
```diff
@@ -993,6 +996,46 @@ class AmplifyAuthCognitoDart extends AuthPluginInterface
...
+  @override
+  Future<CognitoDevice> fetchCurrentDevice() async {
+    final tokens = await stateMachine.getUserPoolTokens();
+    final deviceSecrets = await _deviceRepo.get(tokens.username);
+    final deviceKey = deviceSecrets?.deviceKey;
+    if (deviceSecrets == null || deviceKey == null) {
+      throw const DeviceNotTrackedException();
+    }
+
+    late GetDeviceResponse resp;
```
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart:1008]: nit: IMO the abbreviation doesn't add much value here.
```suggestion
    late GetDeviceResponse response;
```

File: packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart [hunk 2]
```diff
@@ -993,6 +996,46 @@ class AmplifyAuthCognitoDart extends AuthPluginInterface
...
+    final device = resp.device;
+    final attributes =
+        device.deviceAttributes ?? const <cognito.AttributeType>[];
+
+    return CognitoDevice(
+      id: deviceKey,
+      attributes: {
+        for (final attribute in attributes)
+          attribute.name: attribute.value ?? '',
+      },
```
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart:1032]: Q: Does this need a for loop? Does `attributes` have other properties? 
```suggestion
      attributes: attributes,
```

Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart:1032]: > Q: Does this need a for loop? 

No, you can rewrite this (after checking attributes is [not empty](https://pub.dev/documentation/built_collection/latest/built_collection/BuiltList/isNotEmpty.html) like this:
```
    final test = attributes.elementAt(0);
    final name = test.name;
    final value = test.value ?? '';
    ...
    attributes = {name: value}
```
... among other methods in the [BuiltList class](https://pub.dev/documentation/built_collection/latest/built_collection/BuiltList-class.html). The loop is mostly stylistic and for parity, being that this is how the other devices API's extracts the name and values from the attributes map. 

Note that `attributes: attributes` is type invalid.  

> Does `attributes` have other properties?

As of now, no

File: packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart (hunk 1 above)
Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart]: added

File: packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart (hunk 2 above)
Comment (Equartey) [reply packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart:1032]: Makes sense, thanks for adding that context.

**Review by Equartey — APPROVED** [review]
LGTM

**Pull request merged**